### ISSUE TITLE
[REVIEW]: Metatheory.jl: Fast and Elegant Algebraic Computation in Julia with Extensible Equality Saturation

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3078]
**Submitting author:** @ 0x0f0f0f (<a href="http://orcid.org/0000-0002-8122-9469">Alessandro Cheli</a>)
**Repository:** <a href="https://github.com/0x0f0f0f/Metatheory.jl" target ="_blank">https://github.com/0x0f0f0f/Metatheory.jl</a>
**Version:** v0.3.2
**Editor:** @dpsanders
**Reviewers:** @mwillsey, @jpfairbanks, @philzook58
**Archive:** <a href="https://doi.org/10.5281/zenodo.4646136" target="_blank">10.5281/zenodo.4646136</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/3266e8a08a75b9be2f194126a9c6f0e9/status.svg)](https://joss.theoj.org/papers/3266e8a08a75b9be2f194126a9c6f0e9)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/3266e8a08a75b9be2f194126a9c6f0e9"><img src="https://joss.theoj.org/papers/3266e8a08a75b9be2f194126a9c6f0e9/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/3266e8a08a75b9be2f194126a9c6f0e9/status.svg)](https://joss.theoj.org/papers/3266e8a08a75b9be2f194126a9c6f0e9)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/0x0f0f0f/Metatheory.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@mwillsey & @jpfairbanks & @philzook58, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @dpsanders know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/0x0f0f0f/Metatheory.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@ 0x0f0f0f) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/0x0f0f0f/Metatheory.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@ 0x0f0f0f) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/0x0f0f0f/Metatheory.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@ 0x0f0f0f) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @mwillsey, @jpfairbanks, @philzook58 it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.38 s (158.9 files/s, 10624.5 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                           44            574            303           2258
Markdown                         7            145              0            418
TeX                              2             21              0            164
YAML                             4              1             10             83
TOML                             2              4              0             29
SVG                              1              0              0              3
-------------------------------------------------------------------------------
SUM:                            60            745            313           2955
-------------------------------------------------------------------------------


Statistical information for the repository 'ba80bcb2091f1b9ae0a21830' was
gathered on 2021/03/02.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/3434304 is OK
- 10.1137/141000671 is OK
- 10.1145/2813885.2737959 is OK
- 10.1145/2737924.2737959 is OK
- 10.1145/322186.322198 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by 0x0f0f0f** [comment]
I will release 0.1.2 in the next days, can the version be updated here in the review?

**Comment by 0x0f0f0f** [comment]
Thanks. I've tried to change the version number in the pre review but I did not have editor permissions to do so. Please keep in mind that a lot of things changed since v0.1.0 (many bugfixes) and the documentation in the repo is currently referring to the latest commit in the master branch of Metatheory.jl. The general concepts explained in the paper are unaltered.

**Comment by dpsanders** [comment]
There is no version number associated at the review stage. The reviewers will be looking at the `master` branch. That is why it is useful for you to link PRs here so that they can see what is being changed at each step.

Once the review is finished, that finalised version will be the one that is associated with the published paper.

**Comment by dpsanders** [comment]
@0x0f0f0f: I think it would be helpful to include a simple example of the application of the library in the paper.

**Comment by dpsanders** [comment]
@0x0f0f0f: The package will need proper narrative documentation, not just a quick README and the API reference.
See https://documentation.divio.com/ for a discussion of the different types of documentation required by a package.

**Comment by jpfairbanks** [comment]
1. The main figure is a reproduction from the EGG paper. There should be a novel figure highlighting the contributions of this library over and above the EGG library. 
2. The primary benefit of using Metatheory is the algebraic nature of the specification of a simplification system. I would like that to be highlighted in the text.
3. Examples of specifying algebraic theories should be provided including the composition of theories by concatenating their rules.
4. Why is concatenating rules sufficient for composition in this setting? Is it because these are all single-sorted theories with universally quantified equations? 

I agree with David that this package needs additional narrative documentation. Some content in the readme would be appropriate for that documentation.

**Comment by 0x0f0f0f** [comment]
Points

> 1. The main figure is a reproduction from the EGG paper. There should be a novel figure highlighting the contributions of this library over and above the EGG library.
> 2. The primary benefit of using Metatheory is the algebraic nature of the specification of a simplification system. I would like that to be highlighted in the text.
> 3. Examples of specifying algebraic theories should be provided including the composition of theories by concatenating their rules.
> 4. Why is concatenating rules sufficient for composition in this setting? Is it because these are all single-sorted theories with universally quantified equations?

These 4 points should have been added in the paper in the latest commit.

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by mwillsey** [comment]
Low-level feedback
- Line 26: "Theories can then be executed". Theories should be introduced first, this is an abrupt start.
- "rust" should be "Rust" throughout

Overall the paper, software, and documentation (especially) looks good. As the egg author, I view "egg in Julia" as a sufficient contribution, given that Metatheory.jl capitalizes on then dynamic nature of Julia with, e.g. homoiconicity, runtime code generation, and so on.

My only overall comment is that (classical, directed) rewriting deserves a slightly longer treatment. Lots of work has gone into making rewriting fast and effective (e.g. completion, reduction orders, etc.), and using equality saturation will be slower in many use cases; this is simply a tradeoff.

**Comment by 0x0f0f0f** [comment]
Thank you max! I really appreciate your comments and suggestions. I wanted to keep this paper short and introductory (focusing on the important contributions) to leave a more theoretical (and analytical) approach to a follow-up. I'd be really happy to read and write more about classical rewriting. Please send me an email if you have any paper to suggest!

**Comment by mwillsey** [comment]
@0x0f0f0f I really like this survey paper, it discusses most of the seminal works in this area: http://www.math.tau.ac.il/~nachumd/papers/taste-fixed.pdf

**Comment by 0x0f0f0f** [comment]
@jpfairbanks how does the current documentation look? Anything you think I should add?
Thank you Max! Very interesting paper. Well written and full of examples.

**Comment by whedon** [comment]
:wave: @jpfairbanks, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @philzook58, please update us on how your review is going (this is an automated reminder).

**Comment by philzook58** [comment]
I have already found Metatheory.jl to be an extremely useful addition to the space with thoughtful insights in regards to usability, flexibility, and integration with the host language.

I have an intermixing of syntactic/grammar/style comments and actual content questions. I marked these comments as Style, since their suggestions are even more optional. I hope I am not overstepping my purview as a reviewer.

- " a well-known paradigm similar to several LISP idioms such as Scheme." >> I wouldn't call Scheme a "Lisp idiom". I would call it a dialect. Suggestion : "a well-known paradigm found in LISP dialects such as Scheme".  I've had discussions about what exactly homoiconicity is and I'm personally not sure. I'm unclear if it is a "paradigm" persay. But this is very subjective. Wikipedia seems to use the term "property of a programming language" which feels more appropriate to me. https://en.wikipedia.org/wiki/Homoiconicity

- Style - "Julia programming language, designed to take advantage of >> its << powerful reflection capabilities to"

- "Intuitively, Metatheory.jl transforms Julia expressions in other Julia expressions at both compile and run time.". I think this makes sense and I see what your going for, but Julia has a looser notion of compile and run time than most languages. I don't have a good suggestion.

- Style - "Theories can then be executed throught two, highly composable, rewriting backends. The first one, is based just on standard rewriting built on top of the pattern matcher developed in [@matchcore]." I think you do not need any of these commas. There is also a typo "throught"

- Style - "Such an approach suffers of the usual problems of rewriting systems. Even trivial equational rules such as commutativity may lead to non-terminating systems and thus need to be adjusted by some sort of structuring or rewriting order. This requires extensive user reasoning."

- Style - "The other back-end for Metatheory.jl, the core of our contribution, is designed to not require the user to reason about rewriting order by relying on >> equality saturation on e-graphs, the state-of-the-art technique adapted from the egg rust library [@egg]. <<" adjustment of sentence.

-  Provided with a theory of equational rewriting rules, defined in pure Julia, e-graphs compactly represent many equivalent programs.  >>> An egraph can be usefully used without equational pattern matching rewriting rules by merely inserting a fixed number of ground terms and equivalences into it. This is the usual mode in which they are used in quantifier free SMT queries. I feel this may be conflating two distinct aspects of the egraph technique as inseparable.

- "Saturation iteratively executes an e-graph specific pattern matcher to efficiently compute (and analyze) all possible equivalent expressions contained in the e-graph congruence closure." >>> In my understanding, the definition of the egraph includes that it propagates all congruence closures at some point. The term saturation refers to saturation with respect to the rules, not with respect to the congruence closure.

- Style - " beyond being easy to use and composable, " >> I erased "to be"

- Style - I would suggest removing basically all colons ":" throughout the paper and just starting new sentences.

-  "such as costs of nodes or logical statements attached to terms." The logical statements part is unclear to me how it would work. I think this may be referring to discussions we've had so I want to bring this up. Is this an expected thing to be able to do or something already done?

- "the most fitting expressions from an >> equivalence class represented in an e-graph." >> I don't think congruence closure is the appropriate term here.

- Style - "Many applications of equality saturation have been recently published, tailoring advanced optimization tasks." This use of the word "tailoring" feels a little odd to me. Suggestion: "Many applications of equality saturation to advanced optimization tasks have been recently published."

- Style - "In [@yang2021equality], authors used egg to superoptimize tensor signal flow graphs describing neural networks. However, Herbie requires interoperation and conversion of expressions between different languages and libraries." I would switch these sentences. Finish talking about Herbie.

**Comment by mwillsey** [comment]
@philzook58 these are great points, I agree with them. Aside from that, I consider my review complete.

**Comment by philzook58** [comment]
I think I need a new invitation link as I can't edit my checkboxes and it appears to have expired. My apologies 

I would also like to note for the sake of transparency that I have been in private discussion with the author on progress and usage of the library before even this article was submitted (it is mentioned in the acknowledgements of the paper also). We have no other prior connections besides that. I do not consider this to be a true conflict of interest.

**Comment by 0x0f0f0f** [comment]
@philzook58 I don't think it is a true conflict of interest. AFAIK you've been constantly reviewing my implementation decisions even before I've submitted the paper.

**Comment by dpsanders** [comment]
@whedon invite @philzook58 as reviewer

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by dpsanders** [comment]
@whedon add @philzook58 as reviewer

**Comment by whedon** [comment]
OK, @philzook58 is now a reviewer

**Comment by dpsanders** [comment]
@philzook58: I re-invited you; let me know if that didn't work. Thanks!

**Comment by dpsanders** [comment]
@jpfairbanks: Are you done with your review? Could you please fill in the remaining checkboxes at the top of this page? Thanks!

**Comment by 0x0f0f0f** [comment]
Updated the pdf with corrections from Phil

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by jpfairbanks** [comment]
I agree with @philzook58's suggestions. I think that the Metatheory library is rapidly evolving as we review this document, and I am not sure about the editorial standard for JOSS. Are libraries supposed to be mostly mature at the time of JOSS publication? Metatheory.jl is rapidly gaining new features, which might prompt a delay in JOSS publication, but I would leave that decision to the editors including @dpsanders. For example, since the original submission, Metatheory now defines an interface for term like types, which enables support for user supplied types to be used for terms. This is a large improvement over the original EGG implementation and should be called out in this article. I'm not sure what the editorial standard on this is. I am happy to see improvements in the library, but I think we will need a new JOSS submission in a month to document all the new features. Then again, if JOSS entries are living documents that always point to the latest version of the libraries docs, then this is less of a concern.

I also have been in contact with @0x0f0f0f about the integration of typed expressions with the goal of using Metatheory in my own work.

**Comment by 0x0f0f0f** [comment]
I'm planning to write another full length paper in detail explaining all the features! This is an introductory paper to serve as canonical citation in the style of `Flux.jl`. In this paper I want to illustrate the core features and high level architecture of Metatheory.jl (egraphs and classical rewriting), which are unlikely to change.

**Comment by 0x0f0f0f** [comment]
I made a single breaking change and I've updated the example in the paper. Will be present in next commit pushed.

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by 0x0f0f0f** [comment]
Only change is that on the last line of example you do not need anymore to add the analysis explicitly but you can directly call `extract!(egraph, cost_function)`. @philzook58 everything looking ok? Were you able to edit the whedon comment with the list after Daniel re-added you?

**Comment by philzook58** [comment]
Everything looks good. Two more comments:

1. I've been added as a reviewer twice in the paper draft
2.  I suspect the code is not quite displaying right in the draft for me. In particular, i think the \cup occurring in the theory combination is not rendering. Maybe I'm mistaken about the syntax. @0x0f0f0f should double check if there is anything else that looks off there

With the two points addressed, I think I consider my review complete

**Comment by 0x0f0f0f** [comment]
Yep. The cup is not rendering. Will fix with `union(...)` asap!

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by 0x0f0f0f** [comment]
Rendering looks good!! Thank you everybody for all the efforts in reviewing my first paper, for all the help, useful comments and all the long chats we had about this package!

**Comment by dpsanders** [comment]
👋 @philzook58, @mwillsey, @jpfairbanks: Many thanks for your comments. Could you please confirm here that you are happy with the paper and repository as is?

**Comment by jpfairbanks** [comment]
LGTM

**Comment by philzook58** [comment]
Looks good to me. I am still listed as a reviewer twice in the sidebar though. I suspect the author has no control over this and it is something to do with whedon. Perhaps my old invitation needs to be cancelled?

**Comment by 0x0f0f0f** [comment]
Should we open an issue in the whedon repo?

**Comment by xuanxu** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by philzook58** [comment]
Duplication fixed! Looks Good!

**Comment by 0x0f0f0f** [comment]
LGTM! Thanks!

**Comment by dpsanders** [comment]
@0x0f0f0f: Please add DOIs for all the references.

**Comment by dpsanders** [comment]
@0x0f0f0f: Please add an explanation for the meaning of Figure 1. What do the arrows and dashed boxes represent? I suggest you also lightly shade or colour what is new in each subsequent part of the figure.

**Comment by 0x0f0f0f** [comment]
@dpsanders some references are github repos and do not have DOIs. Do I have to remove them from `.bib` file and move the links to the paper body?

**Comment by 0x0f0f0f** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/3434304 is OK
- 10.1137/141000671 is OK
- 10.1145/2813885.2737959 is OK
- 10.1145/2737924.2737959 is OK
- 10.1145/322186.322198 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by 0x0f0f0f** [comment]
@dpsanders added explanation and coloring to the figure

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by 0x0f0f0f** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/3434304 is OK
- 10.1137/141000671 is OK
- 10.1145/3428275 is OK
- 10.1145/3276483 is OK
- 10.1145/2813885.2737959 is OK
- 10.1145/2737924.2737959 is OK
- 10.1145/322186.322198 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by dpsanders** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by dpsanders** [comment]
Could you please make a new tagged release and archive it, e.g. at zenodo.org.
Then please report the version number and archive DOI here. Thanks!

**Comment by 0x0f0f0f** [comment]
10.5281/zenodo.4646136

v0.3.2

**Comment by dpsanders** [comment]
@whedon set v0.3.2 as version

**Comment by whedon** [comment]
OK. v0.3.2 is the version.

**Comment by dpsanders** [comment]
@whedon set 10.5281/zenodo.4646136 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4646136" target="_blank">10.5281/zenodo.4646136</a> is the archive.

**Comment by dpsanders** [comment]
@whedon  accept

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by dpsanders** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# All commands can be run on a non-default branch, to do this pass a custom 
# branch name by following the command with `from branch custom-branch-name`.
# For example:

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a dry run of accepting the paper and depositing with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by dpsanders** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/3434304 is OK
- 10.1137/141000671 is OK
- 10.1145/3428275 is OK
- 10.1145/3276483 is OK
- 10.1145/2813885.2737959 is OK
- 10.1145/2737924.2737959 is OK
- 10.1145/322186.322198 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2164

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2164, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by 0x0f0f0f** [comment]
Thanks everybody!

**Comment by kthyng** [comment]
@0x0f0f0f Please change the metadata in your Zenodo archive so the author and title exactly match your JOSS paper.

**Comment by kthyng** [comment]
@0x0f0f0f Starting to read your paper. I don't see a reason to put your position in your author list ("undergraduate student"). Do you have a special reason to include it? If not I suggest removing it. Also is there is a reason you have an additional label as "independent researcher"? If you are attending University of Pisa, you could just list that, though you should expand the information to include at least the country.

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by 0x0f0f0f** [comment]
@kthyng Thanks! I should have fixed everything.

**Comment by kthyng** [comment]
@0x0f0f0f I wasn't able to finish reading your paper earlier, sorry. But I've looked through it now and just a few more comments:
1. Please differentiate between references that are cited inline, in which case you should remove the [] from the reference, and references that are parenthetical (usually at the end of the sentence) which should keep the [] around the reference. You can get more info [here](https://joss.readthedocs.io/en/latest/submitting.html#example-paper-and-bibliography).
2. At least a couple references in the reference list look like they should have "Julia" capitalized (i think it is supposed to be capitalized?). You can preserve capitalization from the .bib file by placing {} around those words.

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by 0x0f0f0f** [comment]
@kthyng  Thanks again for the formatting suggestions. LGTM now

**Comment by dpsanders** [comment]
You're still missing a capital J in the last reference.

**Comment by 0x0f0f0f** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03078/joss.03078/10.21105.joss.03078.pdf) :page_facing_up: :point_left:

**Comment by 0x0f0f0f** [comment]
Fixed

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1377285297667510274) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2174
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.03078](https://doi.org/10.21105/joss.03078)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kthyng** [comment]
Congrats on your new publication @0x0f0f0f! Thanks to editor @dpsanders and reviewers @mwillsey, @jpfairbanks, and @philzook58 for your expertise, hard work, and time!!

**Comment by dpsanders** [comment]
Congratulations @0x0f0f0f! Many thanks to @mwillsey, @jpfairbanks and @philzook58 for your expert advice and suggestions, and to @kthyng for help with the process.

**Comment by 0x0f0f0f** [comment]
Thanks @mwillsey @jpfairbanks @philzook58 @kthyng for all your useful comments and contributions to this project and this review! It's my first published paper, a great milestone!
